crates/lib3h/src/transport/transport_multiplex/mplex.rs

```diff
@@ -0,0 +1,379 @@
+    // our self channel endpoint
+    endpoint_self: Detach<SelfContextEndpoint<()>>,
+    // ref to our inner transport
+    inner_transport:
+        Detach<TransportActorParentWrapperDyn<TransportMultiplex, MplexToInnerContext>>,
+    // our map of endpoints connecting us to our Routes
+    route_endpoints: Detach<HashMap<LocalRouteSpec, SelfContextEndpoint<()>>>,
+}
+
+impl TransportMultiplex {
+    /// create a new TransportMultiplex Instance
+    pub fn new(inner_transport: DynTransportActor) -> Self {
+        let (endpoint_parent, endpoint_self) = create_ghost_channel();
+        let endpoint_parent = Some(endpoint_parent);
+        let endpoint_self = Detach::new(
+            endpoint_self
+                .as_context_endpoint_builder()
+                .request_id_prefix("mplex_to_parent_")
+                .build(),
+        );
+        let inner_transport = Detach::new(GhostParentWrapperDyn::new(
+            inner_transport,
+            "mplex_to_inner_",
+        ));
+        Self {
+            endpoint_parent,
+            endpoint_self,
+            inner_transport,
+            route_endpoints: Detach::new(HashMap::new()),
+        }
+    }
+
+    /// create a route for a specific agent_id + space_address combination
+    /// we are wrapping a network/machine-level gateway with a machine
+    /// space_address and machineId... the space_address + agent_id parameters
+    /// for this function are the higher-level notions for the AgentSpaceGateway
+    pub fn get_agent_space_route(
+        &mut self,
+        space_address: String,
+        local_agent_id: String,
+    ) -> TransportMultiplexRoute {
+        let (endpoint_parent, endpoint_self) = create_ghost_channel();
+        let endpoint_self = endpoint_self
+            .as_context_endpoint_builder()
+            .request_id_prefix("mplex_to_route_")
+            .build();
+
+        let route_spec = LocalRouteSpec {
+            space_address,
+            local_agent_id,
+        };
+
+        if self
+            .route_endpoints
+            .insert(route_spec.clone(), endpoint_self)
+            .is_some()
+        {
+            panic!("get_agent_space_route can only be called ONCE!");
+        }
+
+        TransportMultiplexRoute::new(route_spec, endpoint_parent)
+    }
+
+    /// The owner of this multiplex (real_engine) has received a DirectMessage
+    /// these at this level are intended to be forwarded up to our routes.
+    /// Collect all the un-packed info that will let us pass it back up the
+    /// chain.
+    pub fn received_data_for_agent_space_route(
+        &mut self,
+        space_address: String,
+        local_agent_id: String,
+        remote_agent_id: String,
+        remote_machine_id: String,
+        unpacked_payload: Vec<u8>,
+    ) {
+        let route_spec = LocalRouteSpec {
+            space_address,
+            local_agent_id,
+        };
+        let address = Url::parse(&format!(
```

Comment:
`address` is a bit confusing since we use it for agent and space address, which implies a hash. How about `uri` or `route`?

Reply:
I've been using address to mean Url, and Id to mean hash. We should totally sync up if we'd like to use different words.

Reply:
ack, I see what you mean... lib3h_protocol::Address is a hash... Ok, I'm going to neet to rewire my own brain to think of that differently... there's probably a lot of places in the transports that I'm using the word address to refer to URLs : (

location? route? ... waypoint?

Reply:
I'm going to go with `path` for now... added DND checkbox in https://github.com/holochain/lib3h/issues/306

Reply:
done